crates/bifrost/src/providers/replicated_loglet/network.rs

```diff
@@ -58,10 +63,164 @@ impl RequestPump {
                 _ = &mut cancel => {
                     break;
                 }
-                Some(_append) = self.append_stream.next() => {
+                Some(append) = self.append_stream.next() => {
+                    self.handle_append(Arc::clone(&provider), append).await;
                 }
             }
         }
+
+        Ok(())
+    }
+
+    /// Infailable handle_append method
+    async fn handle_append<T: TransportConnect>(
+        &mut self,
+        provider: Arc<ReplicatedLogletProvider<T>>,
+        incoming: Incoming<Append>,
+    ) {
+        let task = HandleAppendTask {
+            metadata: self.metadata.clone(),
+            provider,
+        };
+
+        // we spawn immediately to avoid slowing the request pump loop
+        // in case responses will go over a slow channel
+        let _ = task_center().spawn_child(
+            TaskKind::Disposable,
+            "handle-append-request",
+            None,
+            task.run(incoming),
+        );
+    }
+}
+
+struct HandleAppendTask<T> {
+    provider: Arc<ReplicatedLogletProvider<T>>,
+    metadata: Metadata,
+}
+
+impl<T> HandleAppendTask<T>
+where
+    T: TransportConnect,
+{
+    fn get_loglet(
+        &self,
+        provider: &ReplicatedLogletProvider<T>,
+        header: &CommonRequestHeader,
+    ) -> Result<Arc<ReplicatedLoglet<T>>, SequencerStatus> {
+        let logs = self.metadata.logs();
+        let chain = logs
+            .chain(&header.log_id)
+            .ok_or(SequencerStatus::BadLogletId)?;
+
+        // appends are only possible to tail segment.
+        let segment = chain
+            .iter()
+            .rev()
+            .find(|segment| segment.index() == header.segment_index)
+            .ok_or(SequencerStatus::BadLogletId)?;
+
+        let loglet = provider
+            .get_replicated_loglet(header.log_id, header.segment_index, &segment.config.params)
+            .map_err(|err| SequencerStatus::Error(err.to_string()))?;
+
+        Ok(loglet)
+    }
```

Comment:
Can we get the loglet from the provider before we start the task? This way we won't need to clone the provider.